training/src/main/scala/com/airbnb/aerosolve/training/NDTree.scala

```diff
@@ -61,16 +64,33 @@ object NDTree {
     if ((indices.length > 0) && (dimensions > 0) && isConstantDimension) {
       val node = new NDTreeNode()
       nodes.append(node)
-      buildTreeRecursive(options, points, indices, nodes, node, 1, splitType)
+      buildTreeRecursive(options, points, -1, indices, nodes, node, 1, splitType)
     }
 
     val tree = new NDTree(nodes.toArray)
     tree
   }
 
+  def getNextAxis(lastAxis: Int,
+                  deltas: Array[Double],
+                  node: NDTreeNode,
+                  minLeafWidthPercentage: Double):Int = {
+    for(i <- 1 to deltas.length) {
+      val axis = (lastAxis + i) % deltas.length
+      val rootWidth = node.getMax.get(axis) - node.getMin.get(axis)
+      val width = deltas(axis)
+      if (rootWidth > 0 && width > 0 &&
+          minLeafWidthPercentage * rootWidth <= width) {
+        return axis
```

Comment:
IIRC `return` in Scala is more expensive especially in for loop, because Scala translate for loop into `flatMap`. You can try `find` function on a collection instead which returns the first occurrence the statement is true.